File: structural/facade.go
package main

import "fmt"

// Subsystems
type Light struct{}
func (l Light) On() { fmt.Println("Lights ON") }
func (l Light) Off() { fmt.Println("Lights OFF") }

type AC struct{}
func (a AC) On() { fmt.Println("AC ON") }
func (a AC) Off() { fmt.Println("AC OFF") }

type Security struct{}
func (s Security) Arm() { fmt.Println("Security Armed") }
func (s Security) Disarm() { fmt.Println("Security Disarmed") }

// Facade
type SmartHomeFacade struct {
	light    Light
	ac       AC
	security Security
}

func (f SmartHomeFacade) LeaveHome() {
	f.light.Off()
	f.ac.Off()
	f.security.Arm()
}

func (f SmartHomeFacade) ArriveHome() {
	f.security.Disarm()
	f.light.On()
	f.ac.On()
}

func main() {
	home := SmartHomeFacade{}
	home.LeaveHome()
	home.ArriveHome()
}

File: structural/facade.php
<?php

// Subsystems
class Light {
    public function on() { echo "Lights ON\n"; }
    public function off() { echo "Lights OFF\n"; }
}

class AC {
    public function on() { echo "AC ON\n"; }
    public function off() { echo "AC OFF\n"; }
}

class Security {
    public function arm() { echo "Security Armed\n"; }
    public function disarm() { echo "Security Disarmed\n"; }
}

// Facade
class SmartHomeFacade {
    private Light $light;
    private AC $ac;
    private Security $security;

    public function __construct() {
        $this->light = new Light();
        $this->ac = new AC();
        $this->security = new Security();
    }

    public function leaveHome() {
        $this->light->off();
        $this->ac->off();
        $this->security->arm();
    }

    public function arriveHome() {
        $this->security->disarm();
        $this->light->on();
        $this->ac->on();
    }
}

// استفاده
$home = new SmartHomeFacade();
$home->leaveHome();
$home->arriveHome();


//Lights OFF
//AC OFF
//Security Armed
//Security Disarmed
//Lights ON
//AC ON
